.gitmodules

```diff
@@ -17,3 +17,7 @@
 	url = https://github.com/llvm-mirror/compiler-rt
 	branch = release_70
 	ignore = all
+[submodule "3rdparty/openssl/openssl"]
+	path = 3rdparty/openssl/openssl
+	url = https://github.com/openssl/openssl
```

Comment:
This PR ports the openssl as it is so I directly point to the openssl git url. My understanding is that we only need to create a fork under openenclave if we need to modify the project.

Reply:
The way I look at it is that we are making quite a bit of changes to get openssl working within the enclave. 
The entire build system is being rewritten from the scratch to get openssl building with our cmake system.

Reply:
All those changes are outside the actual `openssl` repo so there is a clear separation. Even if we make a fork, I don't think those changes should go into the `openssl` repo. Instead, we will end up introducing another layer of redirection when we currently don't have that model.

Reply:
If we create a fork, I'd expect these changes to go into `OpenSSL_1_1_1-stable-oe` branch. That makes it quite easy to compare `OpenSSL_1_1_1-stable-oe` to ``OpenSSL_1_1_1-stable`, `master` and `upstream` to find out differences easily.

To illustrate, here is the exact set of changes required to make `gperftools-2.7` (tcmalloc) work with enclaves (branch openenclave-port.2.7):
https://github.com/gperftools/gperftools/compare/gperftools-2.7...openenclave:openenclave-port-2.7

Reply:
@anakrish I dont think we should create a fork unless we can upstream those changes.
I remember we discussed this in SIG-Arch earlier. Can we upstream changes? 
@AevaOnline @dthaler @CodeMonkeyLeet  We started this discussion, but did not end up with a conclusion. For now, I am in favor of a submodule and a patch.

Reply:
@radhikaj 
Forks allow better maintainability when we are making important changes (e.g a new cmake based build system as in this PR) to a 3rdparty repository irrespective of whether we will ever upstream or not. What is the drawback of using a fork?

Reply:
@anakrish  @radhikaj 
Currently no patch is required to support OpenSSL (therefore, nothing needs to be upstream) so that's why I think directly using the mainstream repo is fine here.
The cmake script is highly coupled with the OE (e.g., linking against `oe_includes`,  installing to the `openenclave-targets`, ..) so it's difficult to review/track if we put it to the separated repo. That is, one needs to go back and forth between two repos to understand the cmake logic.

Now there is two extra hacks needed by OE:
- Checking-in three auto-generated headers. This is because of the `perl` from `git` does not support a specific perl package that is required by the OpenSSL configuration script.
- Checking-in a set of assembly files in order to ensure the coverage of the LVI mitigation.

Again, there is also nothing to be upstreamed, but it's useful to track those extra files. We can discuss more on how/where should we cache them.

Reply:
FWIW, I agree that there doesn't seem to be an immediate need for a fork, and I don't think we should fork dependency projects for changes without the intention to upstream them. 

There's definitely a need for us to document the integration with OpenSSL and to expand the task list that is docs/DesignDocs/enclave_openssl_support.md into a maintenance reference for all the integration points with a version of openssl that works in OE, but I don't think the bulk of that work belongs in an OpenSSL repo, forked or otherwise:
* Building OpenSSL libraries against OE dependencies with OE build restrictions/recommendations (e.g. LVI)
* Running OpenSSL tests in an OE enclave environment.
* Functional hooking of OpenSSL against OE-specific behavior (e.g. entropy provisioning, which we do as a separate lib linked into OE's mbedtls wrapper) 

As long as we're not actually modifying OpenSSL repo files, the cost of maintaining a separate fork strikes me as being higher as it requires an expansion of CI/CD support to a separate project with its own prerequisites; alternatively, if it only gets tested as a submodule through OE SDK, it's problematic for anyone who ends up seeing the project as potentially standalone and tries to consume it directly.

Reply:
Are we planning to modify OpenSSL at all? My understanding was that we'd soon start modifying OpenSSL.
In that case, I'm not sure if carrying around patch files is the right thing to do.

My biggest concern with not using a fork is that the Open Enclave repository is littered with changes required to port each library (OpenSSL, libgcov, libunwind ...) to work within enclaves, increasing the complexity. To me it is logically cleaner to house changes related to porting OpenSSL in a separate repository (a fork or otherwise (e.g SGX-SSL)).

I'd also like to see the set of changes that'd be required to make OpenSSL work for OPTEE as well (assuming that is something we want to support). Would that significantly patches we need to keep in the Open Enclave repository?